source/codegen/metadata_mutation.py

```diff
@@ -64,7 +64,12 @@ def mark_non_proto_params(parameters):
             if mechanism == 'len':
                 if 'determine_size_from' not in size_param:
                     size_param['determine_size_from'] = []
+                    size_param['linked_params_are_optional'] = True
                 size_param['determine_size_from'].append(param['name'])
+                is_optional = common_helpers.is_optional(param)
+                if is_optional and not size_param['linked_params_are_optional']:
```

Comment:
This won't detect the case when the first linked parameter is optional and the second linked parameter is not optional, right? Maybe it makes more sense to initialize `size_param['linked_params_are_optional']` to `None` above instead of True, then you can do something like

```python
if size_param['linked_params_are_optional'] is not None and size_param['linked_params_are_optional'] != is_optional:
```

Reply:
Yikes, that wasn't even close. I addressed this a little bit differently.

```python
            if mechanism == 'len':
                if 'determine_size_from' not in size_param:
                    size_param['determine_size_from'] = []
                size_param['determine_size_from'].append(param['name'])
                is_optional = common_helpers.is_optional(param)
                if is_optional != size_param.get('linked_params_are_optional', is_optional):
                    raise Exception("Code generator does not support linked params that are a mix of optional and required", size_param)
                size_param['linked_params_are_optional'] = is_optional
```

Basically `size_param.get('linked_params_are_optional', is_optional)` is shorthand for "missing is considered a match".

I also did some manual testing around this and it looked OK.